primitives/avn-common/src/event_discovery.rs

```diff
@@ -0,0 +1,198 @@
+}
+
+impl PartialOrd for DiscoveredEvent {
+    fn partial_cmp(&self, other: &Self) -> Option<scale_info::prelude::cmp::Ordering> {
+        // TODO ensure that the comparison is lowercase.
+        let ord_sig = self.event.event_id.signature.partial_cmp(&other.event.event_id.signature);
+
+        if let Some(core::cmp::Ordering::Equal) = ord_sig {
+            return ord_sig
+        }
+
+        match self.block.partial_cmp(&other.block) {
+            Some(core::cmp::Ordering::Equal) => {},
+            ord => return ord,
+        }
+        ord_sig
+    }
+}
+
+impl Ord for DiscoveredEvent {
+    fn cmp(&self, other: &Self) -> scale_info::prelude::cmp::Ordering {
+        // TODO ensure that the comparison is lowercase.
+        let ord_sig = self.event.event_id.signature.cmp(&other.event.event_id.signature);
+
+        if let core::cmp::Ordering::Equal = ord_sig {
+            return ord_sig
+        }
+
+        match self.block.cmp(&other.block) {
+            core::cmp::Ordering::Equal => {},
+            ord => return ord,
+        }
+        ord_sig
+    }
+}
+
+type EthereumEventsSet = BoundedBTreeSet<DiscoveredEvent, IncomingEventsBatchLimit>;
+#[derive(PartialEq, Eq, Clone, Encode, Decode, Debug, TypeInfo, MaxEncodedLen)]
+pub struct EthereumEventsPartition {
+    range: EthBlockRange,
+    partition: u16,
+    is_last: bool,
+    data: EthereumEventsSet,
+}
+
+impl EthereumEventsPartition {
+    pub fn partition(&self) -> u16 {
+        self.partition
+    }
+
+    pub fn events(&self) -> &EthereumEventsSet {
+        &self.data
+    }
+
+    pub fn range(&self) -> &EthBlockRange {
+        &self.range
+    }
+
+    pub fn is_last(&self) -> bool {
+        self.is_last
+    }
+
+    pub fn id(&self) -> H256 {
+        use sp_io::hashing::blake2_256;
+        blake2_256(&(&self).encode()).into()
+    }
+
+    pub fn new(
+        range: EthBlockRange,
+        partition: u16,
+        is_last: bool,
+        data: EthereumEventsSet,
+    ) -> Self {
+        EthereumEventsPartition { range, partition, is_last, data }
+    }
+}
+
+pub type EventsTypesLimit = ConstU32<20>;
+pub type EthBridgeEventsFilter = BoundedBTreeSet<ValidEvents, EventsTypesLimit>;
+
+pub trait EthereumEventsFilterTrait {
+    fn get_filter() -> EthBridgeEventsFilter;
+}
+
+impl EthereumEventsFilterTrait for () {
+    fn get_filter() -> EthBridgeEventsFilter {
+        Default::default()
+    }
+}
+
+pub fn encode_eth_event_submission_data<AccountId: Encode, Data: Encode>(
+    context: &[u8],
+    account_id: &AccountId,
+    data: Data,
+) -> Vec<u8> {
+    log::debug!(
+        "🪲 Encoding submission data: [ context {:?} - account {:?} - data {:?} ]",
+        context,
+        account_id.encode(),
+        &data.encode()
+    );
+    (context, &account_id, data).encode()
+}
+
+pub mod events_helpers {
+    use super::*;
+    pub extern crate alloc;
+    use alloc::collections::BTreeSet;
+
+    pub fn discovered_eth_events_partition_factory(
```

Comment:
Nit: this doesn't feel like a function name :) maybe it should start with a verb

Reply:
typically you don't use verbs on a factory method as it doesn't make sense. I can move it under a struct that is the factory